File: 01-Programming-Languages/Java/Algorithms/Searching/BinarySearch.java
/*
 * Problem: Find the index of a target value in a sorted array.
 * Approach: Binary search — repeatedly divide the search space in half.
 * Time Complexity: O(log n)
 * Space Complexity: O(1)
 * Example: Input: arr = [1, 3, 5, 7, 9], target = 5 -> Output: 2
 */

public class BinarySearch {
    public static int binarySearch(int[] arr, int target) {
        int left = 0, right = arr.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (arr[mid] == target) return mid;
            if (arr[mid] < target) left = mid + 1;
            else right = mid - 1;
        }
        return -1;
    }

    public static void main(String[] args) {
        int[] arr = {1, 3, 5, 7, 9};
        int target = 5;
        System.out.println(binarySearch(arr, target));
    }
}

File: 01-Programming-Languages/R/Algorithms/Searching/binary_search.R
# Problem: Binary Search
# Description: Find the index of a target value in a sorted array.
#
# Approach:
# - Repeatedly divide the search interval in half
# - Compare target with middle element to narrow the search
#
# Time Complexity: O(log n)
# Space Complexity: O(1)
#
# Example:
# Input: arr = c(1, 3, 5, 7, 9, 11), target = 7
# Output: 4

binary_search <- function(arr, target) {
  left <- 1
  right <- length(arr)
  while (left <= right) {
    mid <- left + (right - left) %/% 2
    if (arr[mid] == target) return(mid)
    if (arr[mid] < target) left <- mid + 1
    else right <- mid - 1
  }
  -1
}

arr <- c(1, 3, 5, 7, 9, 11)
target <- 7
idx <- binary_search(arr, target)
cat("Array:", arr, "\n")
cat("Target:", target, "-> Index:", idx, "\n")
